overlord/devicestate/devicestate_remodel_test.go

```diff
@@ -1975,3 +1978,549 @@ func (s *deviceMgrRemodelSuite) TestRemodelUC20LabelConflictsError(c *C) {
+				Timestamp: new.Timestamp(),
+			},
+			{
+				System:    "0000",
+				Model:     model.Model(),
+				BrandID:   model.BrandID(),
+				Timestamp: model.Timestamp(),
+				Revision:  model.Revision(),
+			},
+		})
+	} else {
+		// however, error is still logged, both to the task and the logger
+		c.Check(strings.Join(setModelTask.Log(), "\n"), Matches, tc.taskLogMatch)
+		c.Check(buf.String(), Matches, tc.logMatch)
+
+		c.Assert(seededSystems, HasLen, 1)
+		c.Check(seededSystems[0].SeedTime.Equal(oldSeededTs), Equals, true)
+		seededSystems[0].SeedTime = time.Time{}
+		c.Check(seededSystems, DeepEquals, []devicestate.SeededSystem{
+			{
+				System:    "0000",
+				Model:     model.Model(),
+				BrandID:   model.BrandID(),
+				Timestamp: model.Timestamp(),
+				Revision:  model.Revision(),
+			},
+		})
+	}
+}
+
+func (s *deviceMgrRemodelSuite) TestUC20RemodelSetModelHappy(c *C) {
+	s.testUC20RemodelSetModel(c, uc20RemodelSetModelTestCase{
+		resealErr: []error{
+			nil, // promote recovery system
+			nil, // device change pre model write
+			nil, // device change post model write
+		},
+	})
+}
+
+func (s *deviceMgrRemodelSuite) TestUC20RemodelSetModelErr(c *C) {
+	s.testUC20RemodelSetModel(c, uc20RemodelSetModelTestCase{
+		resealErr: []error{
+			nil, // promote tried recovery system
+			// keep this comment so that gofmt does not complain
+			fmt.Errorf("mock reseal error"), // device change pre model write
+		},
+		taskLogMatch: `.* cannot complete remodel: \[cannot switch device: mock reseal error\]`,
+		logMatch:     `(?s).* cannot complete remodel: \[cannot switch device: mock reseal error\].`,
+	})
+}
+
+func (s *deviceMgrRemodelSuite) TestUC20RemodelSetModelWithReboot(c *C) {
+	// check that set-model does the right thing even if it is restarted
+	// after an unexpected reboot; this gets complicated as we cannot
+	// panic() at a random place in the task runner, so we set up the state
+	// such that the set-model task completes once and is re-run again
+
+	s.state.Lock()
+	defer s.state.Unlock()
+	s.state.Set("seeded", true)
+	s.state.Set("refresh-privacy-key", "some-privacy-key")
+
+	devicestate.SetBootOkRan(s.mgr, true)
+
+	s.newFakeStore = func(devBE storecontext.DeviceBackend) snapstate.StoreService {
+		return &freshSessionStore{}
+	}
+
+	nopHandler := func(task *state.Task, _ *tomb.Tomb) error {
+		return nil
+	}
+	// mock out tasks that are too annoying to provide the proper mocking for
+	s.o.TaskRunner().AddHandler("fake-download", nopHandler, nil)
+	s.o.TaskRunner().AddHandler("validate-snap", nopHandler, nil)
+	s.o.TaskRunner().AddHandler("fake-install", nopHandler, nil)
+	s.o.TaskRunner().AddHandler("check-snap", nopHandler, nil)
+	s.o.TaskRunner().AddHandler("request-serial", nopHandler, nil)
+	// create recovery system requests are boot, which is not done here
+	s.o.TaskRunner().AddHandler("create-recovery-system", nopHandler, nil)
+	s.o.TaskRunner().AddHandler("finalize-recovery-system", nopHandler, nil)
```

Comment:
Nitpick, maybe a helper to mock all those (they overlap with some of the handlers from an earlier test).